src/spu-schema/src/server/stream_fetch.rs

```diff
@@ -86,10 +95,53 @@ impl Default for SmartStreamKind {
 #[derive(Debug, Clone, Encoder, Decoder)]
 pub enum SmartStreamWasm {
     Raw(Vec<u8>),
+    /// compressed WASM module payload using Gzip
+    #[fluvio(min_version = 14)]
+    Gzip(Vec<u8>),
     // TODO implement named WASM modules once we have a WASM store
     // Url(String),
 }
 
+fn zip(raw: Vec<u8>) -> io::Result<Vec<u8>> {
+    let mut encoder = GzEncoder::new(raw.as_ref(), Compression::default());
+    let mut buffer = Vec::with_capacity(raw.len());
+    encoder.read_to_end(&mut buffer)?;
+    Ok(buffer)
+}
+
+fn unzip(compressed: Vec<u8>) -> io::Result<Vec<u8>> {
+    let mut decoder = GzDecoder::new(compressed.as_ref());
+    let mut buffer = Vec::with_capacity(compressed.len());
+    decoder.read_to_end(&mut buffer)?;
+    Ok(buffer)
+}
+
+impl SmartStreamWasm {
+    /// returns the gzip-compressed WASM module bytes
+    pub fn to_gzip(self) -> io::Result<Self> {
+        Ok(match self {
+            Self::Raw(raw) => Self::Gzip(zip(raw)?),
+            Self::Gzip(_) => self,
+        })
+    }
+
+    /// returns the raw WASM module bytes
+    pub fn to_raw(self) -> io::Result<Self> {
+        Ok(match self {
+            Self::Raw(_) => self,
+            Self::Gzip(gzipped) => Self::Raw(unzip(gzipped)?),
+        })
+    }
+
+    /// get the raw bytes of the WASM module
+    pub fn get_raw(self) -> io::Result<Vec<u8>> {
```

Comment:
I think there is a good opportunity here to use `Cow` to avoid consuming the `SmartStreamWasm` instance. I tested it out and something like this works really nicely:

```rust
// Functions taking &self generally are named to_*
pub fn to_raw(&self) -> io::Result<std::borrow::Cow<[u8]> {
    Ok(match self {
        Self::Raw(raw) => Cow::Borrowed(raw), // No copying needed, return borrowed representation
        Self::Gzip(gzipped) -> Cow::Owned(unzip(gzipped)?), // Unzipping only needs a slice to the zipped data, and returns an owned Vec
    })
}
```

It looks like your caller does eventually want a Vec though, so you can use this at your callsite:

```rust
let wasm: Cow<[u8]> = module.wasm.get_raw()?;
stream_request.wasm_module = wasm.into_owned(); // Converts Cow<[u8]> to Vec<u8>
```

The end result will ultimately be the same, but since this is a `pub fn` I think it's a good idea to take the extra little steps to make the API really nice and Rusty